src/fobnail-attester.c

```diff
@@ -181,314 +87,585 @@ UsefulBuf encode_meta(struct meta_data *meta)
+    if (!coap_get_data(received, &tmp.len, (const uint8_t **)&tmp.ptr)) {
+        coap_log(LOG_EMERG, "Nonce doesn't contain any data\n");
+        return COAP_RESPONSE_OK;
+    }
+
+    pc->buf.ptr = malloc(tmp.len);
+    pc->buf.len = tmp.len;
+    memcpy(pc->buf.ptr, tmp.ptr, tmp.len);
 
-    /* create new context */
-    coap_context_t* coap_context = NULL;
-    if ((coap_context = coap_new_context(NULL)) != NULL) {
-        if (enable_coap_block_mode)
-            /* enable block handling by libcoap */
-            coap_context_set_block_mode(coap_context, COAP_BLOCK_USE_LIBCOAP | COAP_BLOCK_SINGLE_BODY);
+    return COAP_RESPONSE_OK;
+}
+
+static UsefulBufC get_nonce(coap_context_t *ctx, coap_session_t *session)
+{
+    const char path[] = "api/v1/nonce";
+    coap_pdu_t *pdu = NULL;
+    UsefulBufC nonce = NULLUsefulBufC;
+    provisioning_context_t *pc = coap_session_get_app_data(session);
+
+    /* Construct CoAP message */
+    pdu = coap_pdu_init(COAP_MESSAGE_CON,
+                        COAP_REQUEST_CODE_GET,
+                        coap_new_message_id(session),
+                        coap_session_max_pdu_size(session));
+    if (!pdu) {
+        coap_log(LOG_EMERG, "%s: cannot create PDU\n", path);
+        return nonce;
     }
 
-    return coap_context;
+    /* Add a Uri-Path option */
+    split_and_add_uri(pdu, path);
+
+    coap_register_response_handler(ctx, nonce_handler);
+
+    /* Send the PDU (releases memory allocated for PDU) */
+    coap_send(session, pdu);
+
+    /* Wait for response */
+    while (response == 0)
+        coap_io_process(ctx, COAP_IO_WAIT);
+
+    /* Clear flag and unregister handler for further requests */
+    response = 0;
+    coap_register_response_handler(ctx, NULL);
+
+    nonce = UsefulBuf_Const(pc->buf);
+    pc->buf = NULLUsefulBuf;
+
+    return nonce;
 }
 
 /* --------------------------- main ----------------------------- */
 
-#define UNUSED  __attribute__((unused))
-
 int main(int UNUSED argc, char UNUSED *argv[])
 {
-    int result;
-
-    /* signal handling */
-    signal(SIGINT, signal_handler);
-    signal(SIGTERM, signal_handler);
+    coap_context_t  *ctx = NULL;
+    coap_session_t *session = NULL;
+    coap_address_t dst;
+    int result = EXIT_FAILURE;
+    provisioning_context_t pc = {0};
 
     /* TODO: parse CLI arguments if needed */
-    /*result = parse_command_line_arguments(argc, argv, ...) != 0
-    if (result) {
-        // 1 means help message is displayed, -1 means error
-        return (result == 1) ? EXIT_SUCCESS : EXIT_FAILURE;
-    }*/
 
     /* Generate attestation identity key */
     if (init_tpm_keys() != TSS2_RC_SUCCESS) {
         printf("Cannot generate AIK\n");
         return EXIT_FAILURE;
     }
 
-    /* TODO: sign attestation identity key */
+    coap_startup();
 
-    coap_context_t* coap_context = NULL;
-    coap_endpoint_t* coap_endpoint = NULL;
+    /* Set logging level */
+    coap_set_log_level(LOG_DEBUG);
```

Comment:
Ideally this should be controllable through command line or environment variable